Sources/PubNub/EventEngine/Subscribe/SubscribeTransition.swift

```diff
@@ -0,0 +1,377 @@
+    ].compactMap { $0 }
+    
+    return TransitionResult(
+      state: Subscribe.ReceivingState(input: state.input, cursor: cursor),
+      invocations: finalInvocations
+    )
+  }
+}
+
+fileprivate extension SubscribeTransition {
+  func setReceiveReconnectingState(
+    from state: State,
+    error: PubNubError
+  ) -> TransitionResult<State, Invocation> {
+    return TransitionResult(
+      state: Subscribe.ReceiveReconnectingState(
+        input: state.input,
+        cursor: state.cursor,
+        retryAttempt: ((state as? Subscribe.ReceiveReconnectingState)?.retryAttempt ?? -1) + 1,
+        reason: error
+      )
+    )
+  }
+}
+
+fileprivate extension SubscribeTransition {
+  func setReceiveFailedState(
+    from state: State,
+    error: PubNubError
+  ) -> TransitionResult<State, Invocation> {
+    guard let state = state as? Subscribe.ReceiveReconnectingState else {
+      return TransitionResult(state: state)
+    }
+    return TransitionResult(
+      state: Subscribe.ReceiveFailedState(
+        input: state.input,
+        cursor: state.cursor,
+        error: error
+      ), invocations: [
+        .managed(.emitStatus(change: Subscribe.ConnectionStatusChange(
+          oldStatus: state.connectionStatus,
+          newStatus: .disconnectedUnexpectedly,
+          error: error
+        )))
+      ]
+    )
+  }
+}
+
+fileprivate extension SubscribeTransition {
+  func setStoppedState(from state: State) -> TransitionResult<State, Invocation> {
+    let invocations: [EffectInvocation<Invocation>] = [
+      .managed(.emitStatus(change: Subscribe.ConnectionStatusChange(
+        oldStatus: state.connectionStatus,
+        newStatus: .disconnected,
+        error: nil
+      )))
+    ]
+    let handshakeStoppedTransition: TransitionResult<State, Invocation> = TransitionResult(
+      state: Subscribe.HandshakeStoppedState(input: state.input, cursor: state.cursor),
+      invocations: invocations
+    )
+    let receiveStoppedTransition: TransitionResult<State, Invocation> = TransitionResult(
+      state: Subscribe.ReceiveStoppedState(input: state.input, cursor: state.cursor),
+      invocations: invocations
+    )
+    
+    switch state {
+    case is Subscribe.HandshakingState:
+      return handshakeStoppedTransition
+    case is Subscribe.HandshakeReconnectingState:
+      return handshakeStoppedTransition
+    case is Subscribe.ReceivingState:
+      return receiveStoppedTransition
+    case is Subscribe.ReceiveReconnectingState:
+      return receiveStoppedTransition
+    default:
+      return TransitionResult(state: state)
+    }
+  }
+}
+
+fileprivate extension SubscribeTransition {
+  func setUnsubscribedState(from state: State) -> TransitionResult<State, Invocation> {
+    return TransitionResult(
+      state: Subscribe.UnsubscribedState(),
+      invocations: [
+        .managed(.emitStatus(change: Subscribe.ConnectionStatusChange(
+          oldStatus: state.connectionStatus,
+          newStatus: .disconnected,
```

Comment:
This possible only for transitions from `Receiving*` states because we are “connected”. While in `Handshaking*` states we are not connected yet.

Reply:
Here's the same situations like above. I had to support it, because Swift SDK has still `connecting` status 😞